lte/gateway/python/magma/pipelined/app/gy.py

```diff
@@ -229,6 +294,185 @@ def _install_default_flows_if_not_installed(self, datapath,
 
         return remaining_flows
 
+    def _get_rule_match_flow_msgs(self, imsi, ip_addr, apn_ambr, rule):
+        """
+        Get flow msgs to get stats for a particular rule. Flows will match on
+        IMSI, cookie (the rule num), in/out direction
+
+        Args:
+            imsi (string): subscriber to install rule for
+            ip_addr (string): subscriber session ipv4 address
+            apn_ambr (integer): maximum bandwidth for non-GBR EPS bearers
+            rule (PolicyRule): policy rule proto
+        """
+        rule_num = self._rule_mapper.get_or_create_rule_num(rule.id)
+        priority = self.get_of_priority(rule.priority)
+
+        flow_adds = []
+        for flow in rule.flow_list:
+            try:
+                flow_adds.extend(self._get_classify_rule_flow_msgs(
+                    imsi, ip_addr, apn_ambr, flow, rule_num, priority,
+                    rule.qos, rule.hard_timeout, rule.id, rule.app_name,
+                    rule.app_service_type))
+
+            except FlowMatchError as err:  # invalid match
+                self.logger.error(
+                    "Failed to get flow msg '%s' for subscriber %s: %s",
+                    rule.id, imsi, err)
+                raise err
+        return flow_adds
+
+    def _get_classify_rule_flow_msgs(self, imsi, ip_addr, apn_ambr, flow, rule_num,
```

Comment:
This seems like a lot of code duplication, is enforcement stats doing the exact same thing? Could we add the next table as an arg and put this into a shared file?

Reply:
Yes, this was left in purpose as I don't have visibility on the planned work on the enforcement side and whether we want the apps code to be decoupled or not.
all the code is factored now.